src/core/thread/mesh_forwarder.cpp

```diff
@@ -237,6 +237,50 @@ void MeshForwarder::UpdateIndirectMessages(void)
     }
 }
 
+otError MeshForwarder::EvictIndirectMessage(void)
+{
+    otError error = OT_ERROR_NOT_FOUND;
+
+    for (Message *message = mSendQueue.GetHead(); message; message = message->GetNext())
+    {
+        if (!message->IsChildPending())
+        {
+            continue;
+        }
+
+        RemoveMessage(*message);
+        ExitNow(error = OT_ERROR_NONE);
+    }
+
+exit:
+    return error;
+}
+
+void MeshForwarder::RemoveMessage(Message &aMessage)
+{
+    Child *children;
+    uint8_t numChildren;
+
+    children = GetNetif().GetMle().GetChildren(&numChildren);
+
+    for (uint8_t i = 0; i < numChildren; i++)
+    {
+        if (aMessage.GetChildMask(i))
+        {
+            aMessage.ClearChildMask(i);
+            mSourceMatchController.DecrementMessageCount(children[i]);
```

Comment:
I think we may need to check here in the child entry to see if this message is currently being held as `GetIndirectMessage()` and if so clearing it and any other states we hold on to for the indirect message...

Reply:
Checking `MeshForwarder::ScheduleTransmissionTask()` I think setting the `mIndirectMessage` pointer to `NULL` on the child (`SetIndirectMessage()`) will be sufficient.

Reply:
Good point.  However, setting the child's indirect message pointer to `NULL` should be sufficient.  `MeshForwarder::GetIndirectTransmission()` initializes any associated state.

Reply:
Updated.